Kokkos updates breaking nightly gcc/8.3.0, gcc/9, gcc/10 builds: Kokkos_HostSpace.hpp:312:12: error: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type

Kokkos nightly builds with gcc/8.3.0 and cuda/11.1+gcc/8.3.0 failed with error message

```
18:19:45 /data/jenkins-new/workspace/kokkos_semsrhel7gpu01_cuda111_gcc830/kokkos/core/src/Kokkos_HostSpace.hpp:312:12: error: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘using value_type = using value_type = using value_type = using value_type = class Kokkos::complex<double>’ {aka ‘class Kokkos::complex<double>’}; use assignment or value-initialization instead [-Werror=class-memaccess]
18:19:45      std::memset(dst.data(), 0, sizeof(ValueType) * dst.size());
18:19:45      ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
18:19:45 /data/jenkins-new/workspace/kokkos_semsrhel7gpu01_cuda111_gcc830/kokkos/core/src/Kokkos_Complex.hpp:67:34: note: ‘using value_type = using value_type = using value_type = using value_type = class Kokkos::complex<double>’ {aka ‘class Kokkos::complex<double>’} declared here
18:19:45      alignas(2 * sizeof(RealType))
```

Likely following from merge of PR #4226 , @masterleinad are you able to take a look?
Issue #4237 may be related.

Reproducer instructions (cuda build):
```
module load sems-env sems-cmake/3.17.1 sems-cuda/11.1 sems-gcc/8.3.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,Serial --arch=Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --with-cuda-options=enable_lambda  --no-examples --with-options=enable_large_mem_tests
```

--- a/core/src/impl/Kokkos_ViewMapping.hpp
+++ b/core/src/impl/Kokkos_ViewMapping.hpp
@@ -2849,24 +2849,22 @@ struct ViewValueFunctor;
 
 template <class DeviceType, class ValueType>
 struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
-  using ExecSpace  = typename DeviceType::execution_space;
-  using PolicyType = Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>>;
+  using ExecSpace = typename DeviceType::execution_space;
 
   ExecSpace space;
   ValueType* ptr;
   size_t n;
-  bool destroy;
   std::string name;
 
+  struct ConstructTag {};
+  struct DestroyTag {};
+
   KOKKOS_INLINE_FUNCTION
-  void operator()(const size_t i) const {
-    if (destroy) {
-      (ptr + i)->~ValueType();
-    }  // KOKKOS_IMPL_CUDA_CLANG_WORKAROUND this line causes ptax error
-       // __cxa_begin_catch in nested_view unit-test
-    else {
-      new (ptr + i) ValueType();
-    }
+  void operator()(DestroyTag, const size_t i) const { (ptr + i)->~ValueType(); }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(ConstructTag, const size_t i) const {
+    new (ptr + i) ValueType();
   }
 
   ViewValueFunctor()                        = default;
@@ -2875,29 +2873,37 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
 
   ViewValueFunctor(ExecSpace const& arg_space, ValueType* const arg_ptr,
                    size_t const arg_n, std::string arg_name)
-      : space(arg_space),
-        ptr(arg_ptr),
-        n(arg_n),
-        destroy(false),
-        name(std::move(arg_name)) {}
+      : space(arg_space), ptr(arg_ptr), n(arg_n), name(std::move(arg_name)) {}
 
-  void execute(bool arg) {
-    destroy = arg;
-
-    // Shortcut for zero initialization
+  template <typename Dummy = ValueType>
+  std::enable_if_t<std::is_trivial<Dummy>::value &&
+                   std::is_trivially_copy_assignable<ValueType>::value>
+  construct_dispatch() {
     ValueType value{};
-    if (std::is_trivial<ValueType>::value &&
-        std::is_trivially_copy_assignable<ValueType>::value &&
-        Impl::is_zero_byte(value) && !destroy) {
+    if (Impl::is_zero_byte(value)) {
       (void)ZeroMemset<ExecSpace, ValueType*, typename DeviceType::memory_space,
                        Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
           space,
           Kokkos::View<ValueType*, typename DeviceType::memory_space,
                        Kokkos::MemoryTraits<Kokkos::Unmanaged>>(ptr, n),
           value);
-      return;
+    } else {
+      parallel_for_implementation</*destroy*/ false>();
     }
+  }
 
+  template <typename Dummy = ValueType>
+  std::enable_if_t<!(std::is_trivial<Dummy>::value &&
+                     std::is_trivially_copy_assignable<ValueType>::value)>
+  construct_dispatch() {
+    parallel_for_implementation</*destroy*/ false>();
+  }
+
+  template <bool destroy>
+  void parallel_for_implementation() {
+    using TagType = std::conditional_t<destroy, DestroyTag, ConstructTag>;
+    using PolicyType =
+        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>, TagType>;
     PolicyType policy(0, n);
     std::string functor_name;
     if (!space.in_parallel()) {
@@ -2924,13 +2930,16 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
                                               kpID);
       }
     } else {
-      for (size_t i = 0; i < n; ++i) operator()(i);
+      TagType tag;
+      for (size_t i = 0; i < n; ++i) operator()(tag, i);
     }
   }
 
-  void construct_shared_allocation() { execute(false); }
+  void construct_shared_allocation() { construct_dispatch(); }
 
-  void destroy_shared_allocation() { execute(true); }
+  void destroy_shared_allocation() {
+    parallel_for_implementation</*destroy*/ true>();
+  }
 };
 
 template <class DeviceType, class ValueType>
@@ -2954,21 +2963,32 @@ struct ViewValueFunctor<DeviceType, ValueType, true /* is_scalar */> {
                    size_t const arg_n, std::string arg_name)
       : space(arg_space), ptr(arg_ptr), n(arg_n), name(std::move(arg_name)) {}
 
-  void construct_shared_allocation() {
+  template <typename Dummy = ValueType>
+  std::enable_if_t<std::is_trivial<Dummy>::value &&
+                   std::is_trivially_copy_assignable<Dummy>::value>
+  construct_shared_allocation() {
     // Shortcut for zero initialization
     ValueType value{};
-    if (std::is_trivial<ValueType>::value &&
-        std::is_trivially_copy_assignable<ValueType>::value &&
-        Impl::is_zero_byte(value)) {
+    if (Impl::is_zero_byte(value)) {
       (void)ZeroMemset<ExecSpace, ValueType*, typename DeviceType::memory_space,
                        Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
           space,
           Kokkos::View<ValueType*, typename DeviceType::memory_space,
                        Kokkos::MemoryTraits<Kokkos::Unmanaged>>(ptr, n),
           value);
-      return;
+    } else {
+      parallel_for_implementation();
     }
+  }
 
+  template <typename Dummy = ValueType>
+  std::enable_if_t<!(std::is_trivial<Dummy>::value &&
+                     std::is_trivially_copy_assignable<Dummy>::value)>
+  construct_shared_allocation() {
+    parallel_for_implementation();
+  }
+
+  void parallel_for_implementation() {
     if (!space.in_parallel()) {
       uint64_t kpID = 0;
       if (Kokkos::Profiling::profileLibraryLoaded()) {

--- a/core/src/impl/Kokkos_ViewMapping.hpp
+++ b/core/src/impl/Kokkos_ViewMapping.hpp
@@ -2849,22 +2849,24 @@ struct ViewValueFunctor;
 
 template <class DeviceType, class ValueType>
 struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
-  using ExecSpace = typename DeviceType::execution_space;
+  using ExecSpace  = typename DeviceType::execution_space;
+  using PolicyType = Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>>;
 
   ExecSpace space;
   ValueType* ptr;
   size_t n;
+  bool destroy;
   std::string name;
 
-  struct ConstructTag {};
-  struct DestroyTag {};
-
   KOKKOS_INLINE_FUNCTION
-  void operator()(DestroyTag, const size_t i) const { (ptr + i)->~ValueType(); }
-
-  KOKKOS_INLINE_FUNCTION
-  void operator()(ConstructTag, const size_t i) const {
-    new (ptr + i) ValueType();
+  void operator()(const size_t i) const {
+    if (destroy) {
+      (ptr + i)->~ValueType();
+    }  // KOKKOS_IMPL_CUDA_CLANG_WORKAROUND this line causes ptax error
+       // __cxa_begin_catch in nested_view unit-test
+    else {
+      new (ptr + i) ValueType();
+    }
   }
 
   ViewValueFunctor()                        = default;
@@ -2873,7 +2875,11 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
 
   ViewValueFunctor(ExecSpace const& arg_space, ValueType* const arg_ptr,
                    size_t const arg_n, std::string arg_name)
-      : space(arg_space), ptr(arg_ptr), n(arg_n), name(std::move(arg_name)) {}
+      : space(arg_space),
+        ptr(arg_ptr),
+        n(arg_n),
+        destroy(false),
+        name(std::move(arg_name)) {}
 
   template <typename Dummy = ValueType>
   std::enable_if_t<std::is_trivial<Dummy>::value &&
@@ -2888,7 +2894,7 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
                        Kokkos::MemoryTraits<Kokkos::Unmanaged>>(ptr, n),
           value);
     } else {
-      parallel_for_implementation</*destroy*/ false>();
+      parallel_for_implementation(false);
     }
   }
 
@@ -2896,14 +2902,11 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
   std::enable_if_t<!(std::is_trivial<Dummy>::value &&
                      std::is_trivially_copy_assignable<ValueType>::value)>
   construct_dispatch() {
-    parallel_for_implementation</*destroy*/ false>();
+    parallel_for_implementation(false);
   }
 
-  template <bool destroy>
-  void parallel_for_implementation() {
-    using TagType = std::conditional_t<destroy, DestroyTag, ConstructTag>;
-    using PolicyType =
-        Kokkos::RangePolicy<ExecSpace, Kokkos::IndexType<int64_t>, TagType>;
+  void parallel_for_implementation(bool arg) {
+    destroy = arg;
     PolicyType policy(0, n);
     std::string functor_name;
     if (!space.in_parallel()) {
@@ -2930,16 +2933,13 @@ struct ViewValueFunctor<DeviceType, ValueType, false /* is_scalar */> {
                                               kpID);
       }
     } else {
-      TagType tag;
-      for (size_t i = 0; i < n; ++i) operator()(tag, i);
+      for (size_t i = 0; i < n; ++i) operator()(i);
     }
   }
 
   void construct_shared_allocation() { construct_dispatch(); }
 
-  void destroy_shared_allocation() {
-    parallel_for_implementation</*destroy*/ true>();
-  }
+  void destroy_shared_allocation() { parallel_for_implementation(true); }
 };
 
 template <class DeviceType, class ValueType>

